aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -2223,67 +2223,34 @@ Tensor chain_matmul(TensorList matrices) {
-  /*
-  We can obtain the kron result using tensordot or einsum. The implementation below uses tensordot.
-  In einsum notation suppose we have `self` with dim 4 and `other` with dim 2
-  the result of below tensordot is in einsum 0123, 45 -> 012345.
-  To obtain the correct kron we need to permute and reshape the array.
-  The permutation rule is the following: going from right to left
-  take axes in turn to form the permutation
-  with our example the correct permutation is 012435 and
-  the kron shape is (shape_self[0], shape_self[1], shape_self[3]*shape_other[0],
-  shape_self[4]*shape_other[1])
-  */
-  std::vector<int64_t> self_sizes = self.sizes().vec();
-  std::vector<int64_t> other_sizes = other.sizes().vec();
-  int64_t self_ndim = self.dim();
-  int64_t other_ndim = other.dim();
-  int64_t min_ndim = std::min(self_ndim, other_ndim);
-  int64_t ndim_diff = std::abs(self_ndim - other_ndim);
-
-  std::vector<int64_t> a_axes(self_ndim);
-  std::vector<int64_t> b_axes(other_ndim);
-  std::iota(a_axes.begin(), a_axes.end(), 0);
-  std::iota(b_axes.begin(), b_axes.end(), 0 + self_ndim);
-
-  bool is_a_larger = self_ndim >= other_ndim;
-  std::vector<int64_t> kron_permutation(self_ndim + other_ndim);
-  for (int64_t i = 0; i < ndim_diff; i++) {
-    kron_permutation[i] = is_a_larger ? a_axes[i] : b_axes[i];
-  }
-  for (int64_t i = 0, j = 0; i < min_ndim; i++, j += 2) {
-    kron_permutation[self_ndim + other_ndim - 1 - j] = b_axes[other_ndim - 1 - i];
-    kron_permutation[self_ndim + other_ndim - 1 - j - 1] = a_axes[self_ndim - 1 - i];
-  }
-
-  std::vector<int64_t> result_shape(std::max(self_ndim, other_ndim));
-  for (int64_t i = 0; i < ndim_diff; i++) {
-    result_shape[i] = is_a_larger ? self_sizes[i] : other_sizes[i];
+Tensor& kron_out(Tensor& result, const Tensor& self, const Tensor& other) {
+  auto maxdim = std::max(self.dim(), other.dim());
+  auto pad_self = maxdim - self.dim();
+  auto pad_other = maxdim - other.dim();
+  c10::SmallVector<int64_t, 10> a_reshape(2 * maxdim);
+  c10::SmallVector<int64_t, 10> b_reshape(2 * maxdim);
+  c10::SmallVector<int64_t, 10> result_reshape(maxdim);
+  for (int i = 0; i < maxdim; i++) {
+    a_reshape[2 * i] = i >= pad_self ? self.sizes()[i - pad_self] : 1;
+    a_reshape[2 * i + 1] = 1;
+    b_reshape[2 * i] = 1;
+    b_reshape[2 * i + 1] = i >= pad_other ? other.sizes()[i - pad_other] : 1;
+    result_reshape[i] = a_reshape[2 * i] * b_reshape[2 * i + 1];
   }
-  for (int64_t i = 0; i < min_ndim; i++) {
-    result_shape[ndim_diff + i] = is_a_larger
-        ? self_sizes[ndim_diff + i] * other_sizes[i]
-        : other_sizes[ndim_diff + i] * self_sizes[i];
+  auto self_view = at::_unsafe_view(self, a_reshape);
+  auto other_view = at::_unsafe_view(other, b_reshape);
+  if (!result.defined()) {
+    result = at::_unsafe_view(at::mul(self_view, other_view), result_reshape);
+  } else {
+    at::mul_out(result, self_view, other_view);
+    result.resize_(result_reshape);
```

Comment:
Does this need to use resize_output? Actually, in this case things seem trickier than that.

What kind of tensor can a use pass to out= here without this complaining? If a user passes a tensor with shape `result_reshape` then can `mul_out` yell that it received a tensor of the incorrect size? If a user passes a tensor of the shape that `mul_out` likes then can it be resized without warning? That'd be bad if out= was a view, for example.

Reply:
Is `out=view` a defined behavior?

Reply:
Yes, typically it uses `resize_` on `out`, so if `out` is correct size, it is directly used and view relationship persists, if it is not then anything could happen, so in that sense it's not defined. Here though preserving this behavior would require first re-viewing `out` with expected size of `mul` if possible (for reductions with similar problems I inserted unsqueezes and not cared about actual sizes, if they are wrong the `resize_` would handle it), and then re-viewing it back.

Reply:
Looks like the current approach preserve storage, but will erase discontinuity:
```python
>>> import torch

>>> a = torch.randn(4,6,8,10).contiguous(memory_format=torch.channels_last)
>>> a.stride()
(480, 1, 60, 6)
>>> a.view(2,2,2,3,2,4,2,5).stride()
(960, 480, 3, 1, 240, 60, 30, 6)
>>> a.resize_(2,2,2,3,2,4,2,5).stride()
(960, 480, 240, 80, 40, 10, 5, 1)

>>> a = torch.randn(4,6,8,10).contiguous(memory_format=torch.channels_last)
>>> a.view(2,2,2,3,2,4,2,5).resize_(2,2,2,3,2,4,2,5).stride()
(960, 480, 3, 1, 240, 60, 30, 6)

>>> a = torch.randn(4,6,8,10).contiguous(memory_format=torch.channels_last)
>>> a.storage().data_ptr()
94728832336256
>>> a.resize_(2,2,2,3,2,4,2,5).storage().data_ptr()
94728832336256
```

Reply:
```python
>>> a = torch.randn(2,3,4,5).contiguous(memory_format=torch.channels_last)
>>> b = torch.kron(a, a)
>>> b.stride()
(3600, 1, 225, 9)
>>> torch.kron(a, a, out=b).stride()
(3600, 400, 25, 1)
```

Reply:
Do you happen to have old build? I suspect the behavior was wrong there too. Colab doesn't have kron at all, and my local build includes this PR :-)

Reply:
Ah, no, old approach was always doing a copy in `out`, so it was functionally correct but slow.

Reply:
For old build it is
```python
>>> import torch
>>> a = torch.randn(2,3,4,5).contiguous(memory_format=torch.channels_last)
>>> b = torch.kron(a, a)
>>> b.stride()
(3600, 400, 25, 1)
```

Reply:
Are there other options that copying to it?

Reply:
I think
```C++
    result.resize_(result_reshape);
    at::mul_out(result.view(...), self_view, other_view);
```
will do the work

Reply:
Should the first line use resize_output to throw a warning if appropriate?

Reply:
numpy also doesn't preserve permutation here, like older pytorch build.

Reply:
Should we do 
```
result.resize_(result_reshape, self.suggest_memory_format());
```
instead of `resize_output`?

Reply:
Sorry, why? We want to be consistent with PyTorch, where the out= semantics are the the operation is performed and then its output is "safe copied" to the tensor passed to out=. If the tensor has to be altered then a warning is thrown, unless the tensor had no elements.

Reply:
No, we should preserve any permutation, like we automatically do now, not just channels last. 
For a warning, we can do result.view(...with inserted 1 dimensions...) and rely on mul_out to throw a warning if sizes are wrong, or do as Xiang suggested a few comments up, and use resize_output there instead of resize_, it doesn't really matter.

Reply:
Or maybe add an argument `c10::optional<MemoryFormat> optional_memory_format` to `resize_output`

Reply:
Please no, use of `memory_format` argument is wrong in most cases outside of torch.nn.